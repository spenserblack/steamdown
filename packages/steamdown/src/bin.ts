import { readFileSync } from 'fs';
import { createInterface } from 'readline';
import parse from './index';

const inputArg = process.argv[2];

if (inputArg && inputArg !== '-') {
  const input = readFileSync(inputArg, 'utf8');
  console.log(parse(input));
} else {
  process.stdin.on('data', (data) => {
    console.log(parse(data.toString()));
  });
}
