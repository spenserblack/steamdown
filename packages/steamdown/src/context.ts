/**
 * Context (items that aren't rendered and shouldn't be part of the syntax tree).
 */
export class Context {
  private links = new Map<string, string>();

  public addLink(id: string, url: string) {
    this.links.set(id, url);
  }

  public getLink(id: string) {
    return this.links.get(id);
  }
}
