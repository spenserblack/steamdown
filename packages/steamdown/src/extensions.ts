import type {
  Extension,
  GenericToken,
  MarkedToken,
  StrongToken,
  HrToken,
} from "./types";

export type SpoilerToken = {
  type: "spoiler";
  raw: string;
  text: string;
  tokens: MarkedToken[];
};

export type NoparseToken = {
  type: "noparse";
  raw: string;
  text: string;
};

export type NoparsespanToken = {
  type: "noparsespan";
  raw: string;
  text: string;
};

export type BlockquoteToken = {
  type: "blockquote";
  raw: string;
  tokens: MarkedToken[];
  author?: string | null | undefined;
};

const wrapped = (s: string, wrapper: string): boolean =>
  s.startsWith(wrapper) && s.endsWith(wrapper);

const space = {
  name: "space",
  renderer({ raw }: GenericToken) {
    return raw;
  },
};
const text = {
  name: "text",
  renderer({ raw }: GenericToken) {
    return raw;
  },
};

const underline = {
  name: "strong",
  renderer(token: GenericToken) {
    const { raw, text } = token as StrongToken;
    if (wrapped(raw, "__")) {
      return `[u]${text}[/u]`;
    } else if (wrapped(raw, "**")) {
      return `[b]${text}[/b]`;
    }
    return false;
  },
};

const spoiler: Extension = {
  name: "spoiler",
  level: "inline",
  start(src: string): number {
    // TODO Remove this hack once @types package is updated
    return src.match(/!!!/)?.index as number;
  },
  tokenizer(src: string): SpoilerToken | void {
    const match = src.match(/^!!!([^\n]*)!!!/);
    if (match) {
      return {
        type: "spoiler",
        raw: match[0],
        text: match[1],
        tokens: this.lexer.inlineTokens(match[1]),
      };
    }
  },
  renderer(token: GenericToken) {
    const { tokens } = token as SpoilerToken;
    return `[spoiler]${this.parser.parseInline(tokens)}[/spoiler]`;
  },
};

const noparse: Extension = {
  name: "noparse",
  level: "block",
  start(src: string): number {
    return src.match(/\{\{\{\n/)?.index as number;
  },
  tokenizer(src: string): NoparseToken | void {
    const match = src.match(/^\{\{\{\n((?:.|\n)*\n)\}\}\}\n?/);
    if (match) {
      return {
        type: "noparse",
        raw: match[0],
        text: match[1],
      };
    }
  },
  renderer(token: GenericToken) {
    const { text } = token as NoparseToken;
    return `[noparse]\n${text}[/noparse]\n\n`;
  },
};

const noparsespan: Extension = {
  name: "noparsespan",
  level: "inline",
  start(src: string): number {
    // TODO And this too
    return src.match(/\{\{\{[^\n]/)?.index as number;
  },
  tokenizer(src: string): NoparsespanToken | void {
    const match = src.match(/^\{\{\{(.*)\}\}\}/);
    if (match) {
      return {
        type: "noparsespan",
        raw: match[0],
        text: match[1],
      };
    }
  },
  renderer(token: GenericToken) {
    const { text } = token as NoparsespanToken;
    return `[noparse]${text}[/noparse]`;
  },
};

const hr = {
  name: "hr",
  renderer() {
    return "[hr][/hr]\n\n";
  },
};

const blockquote: Extension = {
  name: "blockquote",
  level: "block",
  start(src: string): number {
    return src.match(/(?:^|\n)> ?.*(?:\n|$)/)?.index as number;
  },
  tokenizer(src: string): BlockquoteToken | void {
    const match = src.match(/^((?:> ?.*(?:\n|$))+)(?:\((.*)\)(?:\n|$))?/);
    if (match) {
      const [raw, text, author] = match as [string, string, string | undefined];
      const tokens = this.lexer.blockTokens(
        text.replace(/^> ?/gm, ""),
        undefined as any,
      );
      return {
        type: "blockquote",
        raw,
        author,
        tokens,
      };
    }
  },
  renderer(token: GenericToken) {
    const { tokens, author } = token as BlockquoteToken;
    const parsed = this.parser.parse(tokens);
    return `[quote${author ? `=${author}` : ""}]${parsed}[/quote]\n\n`;
  },
};

const extensions: Extension[] = [
  space,
  // text,
  underline,
  spoiler,
  noparse,
  noparsespan,
  hr,
  blockquote,
];

export default extensions;
