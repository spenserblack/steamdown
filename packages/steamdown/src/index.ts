import remarkParse from 'remark-parse';
import remarkGfm from "remark-gfm";
import { unified } from "unified";

const processer = unified().use(remarkParse).use(remarkGfm);

/**
 * Renders the given Markdown string to Steam's text format.
 */
export default async function parse(input: string): Promise<string> {
  const processed = await processer.process(input);
  return String(processed);
}
