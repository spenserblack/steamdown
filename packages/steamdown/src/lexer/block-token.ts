import Token from "./token";

export default abstract class BlockToken extends Token {
  public readonly scope = "block";
}
