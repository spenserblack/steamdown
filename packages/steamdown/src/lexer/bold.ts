import InlineToken from "./inline-token";
import TaggedInlineToken from "./tagged-inline-token";

export default class Bold extends TaggedInlineToken {
  public readonly tag = "b";
  public readonly tokens: InlineToken[];

  private constructor(public readonly text: string, literal: string) {
    super(literal);
    this.tokens = InlineToken.lexTokens(text);
  }

  public static hint(md: string): boolean {
    return md.startsWith("**");
  }

  public static lex(md: string): [token: Bold, remainder: string] | null {
    const scan = InlineToken.scan(md.slice(2), "\\*\\*");
    if (!scan) {
      return null;
    }
    const [index, text] = scan;
    // NOTE Off by 2 because scan starts at index 2, regex index
    //      is the start of the match (so the first valid character before
    //      the delimiter), and 2 more to account for the delimiter itself.
    const literal = md.slice(0, index + 5);
    const remainder = md.slice(index + 5);
    return [new Bold(text, literal), remainder];
  }
}
