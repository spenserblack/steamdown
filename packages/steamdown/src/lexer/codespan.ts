import InlineToken from "./inline-token";
import TaggedInlineToken from "./tagged-inline-token";

export default class Codespan extends TaggedInlineToken {
  public readonly tag = "code";
  public readonly tokens: InlineToken[];
  private constructor(public readonly text: string, literal: string) {
    super(literal);
    this.tokens = InlineToken.lexTokens(text);
  }

  public static hint(md: string): boolean {
    return md.startsWith("`");
  }

  public static lex(md: string): [token: Codespan, remainder: string] | null {
    const escapedDelim = "\\`";
    const scan = InlineToken.scan(md.slice(1), escapedDelim);
    if (!scan) {
      return null;
    }
    const [index, text] = scan;
    // NOTE Off by 1 because scan starts at index 1, regex index
    //       is the start of the match (so the first valid character before
    //       the delimiter), and 1 more to account for the delimiter itself.
    const literal = md.slice(0, index + 3);
    const remainder = md.slice(index + 3);
    return [new Codespan(text, literal), remainder];
  }
}
