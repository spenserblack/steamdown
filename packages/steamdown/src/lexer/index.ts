import Token from "./token";
import Heading from "./heading";
import { LexError } from "./errors";
import type { Lexer } from "./token";
export { Bold, Italic, Text, lexInline } from "./inline-tokens";

export { Token, Heading, LexError };

const lexers: Lexer[] = [Heading];
function reduceLexers(md: string): [token: Token, remainder: string] {
  const hintedLexers = lexers.filter((lexer) => lexer.hint(md));
  for (let i = 0; i < hintedLexers.length; ++i) {
    const lexer = hintedLexers[i];
    const lexed = lexer.lex(md);
    if (lexed) {
      return lexed;
    }
  }
  throw new LexError("Could not lex to any value");
}

export default function lex(md: string): Token[] {
  let tokens: Token[] = [];
  let toLex: string = md;
  while (toLex.trimEnd() !== "") {
    const [token, remainder] = reduceLexers(toLex);
    tokens.push(token);
    toLex = remainder;
  }
  return tokens;
}
