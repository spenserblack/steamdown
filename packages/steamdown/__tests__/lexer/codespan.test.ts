import { describe, expect, test } from "@jest/globals";
import { Codespan } from "../../src/lexer";

describe("Codespan", () => {
  test.each([
    ["`foo`", true],
    ["foo", false],
  ])('.hint("%s")', (md, expected) => {
    expect(Codespan.hint(md)).toBe(expected);
  });

  describe.each([
    ["`foo`", "`foo`", "foo", ""],
    ["`a\nb`", "`a\nb`", "a\nb", ""],
    ["`foo` bar", "`foo`", "foo", " bar"],
    ["`foo \\` bar`", "`foo \\` bar`", "foo \\` bar", ""],
  ])('lex("%s")', (md, expectedLiteral, expectedText, expectedRemainder) => {
    test("remainder", () => {
      const result = Codespan.lex(md);
      expect(result).not.toBeNull();
      const [, remainder] = result!;
      expect(remainder).toBe(expectedRemainder);
    });

    test(".literal", () => {
      const result = Codespan.lex(md);
      expect(result).not.toBeNull();
      const [token] = result!;
      expect(token.literal).toBe(expectedLiteral);
    });

    test(".text", () => {
      const result = Codespan.lex(md);
      expect(result).not.toBeNull();
      const [token] = result!;
      expect(token.text).toBe(expectedText);
    });

    test(".tokens", () => {
      const result = Codespan.lex(md);
      expect(result).not.toBeNull();
      const [token] = result!;
      expect(token.tokens).toMatchSnapshot();
    });

    test(".render()", () => {
      const result = Codespan.lex(md);
      expect(result).not.toBeNull();
      const [token] = result!;
      expect(token.render()).toMatchSnapshot();
    });
  });

  test.each([["`foo"], ["`foo`bar"]])('lex("%s")', (md) => {
    expect(Codespan.lex(md)).toBeNull();
  });
});
