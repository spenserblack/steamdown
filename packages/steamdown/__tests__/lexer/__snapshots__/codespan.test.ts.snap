// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Codespan lex("\`a
b\`") .render() 1`] = `
"[code]a
b[/code]"
`;

exports[`Codespan lex("\`a
b\`") .tokens 1`] = `
[
  Text {
    "content": "a
b",
    "literal": "a
b",
    "scope": "inline",
  },
]
`;

exports[`Codespan lex("\`foo \\\` bar\`") .render() 1`] = `"[code]foo \` bar[/code]"`;

exports[`Codespan lex("\`foo \\\` bar\`") .tokens 1`] = `
[
  Text {
    "content": "foo \\\` bar",
    "literal": "foo \\\` bar",
    "scope": "inline",
  },
]
`;

exports[`Codespan lex("\`foo\` bar") .render() 1`] = `"[code]foo[/code]"`;

exports[`Codespan lex("\`foo\` bar") .tokens 1`] = `
[
  Text {
    "content": "foo",
    "literal": "foo",
    "scope": "inline",
  },
]
`;

exports[`Codespan lex("\`foo\`") .render() 1`] = `"[code]foo[/code]"`;

exports[`Codespan lex("\`foo\`") .tokens 1`] = `
[
  Text {
    "content": "foo",
    "literal": "foo",
    "scope": "inline",
  },
]
`;
